[c10d] Separate monitoring thread into a class in PGNCCL

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #153977

This is the start of a series of efforts to consolidating auxiliary threads in PGNCCL, aka watchdog and heartbeat_monitoring threads. Right now we launch these two threads per PG instances, i.e., if users create hundred or thousand instances of PG or subPGs, we will end up with that twice many side threads which is not efficient. We have a RFC to consolidate them (https://github.com/pytorch/pytorch/issues/146956). Right now both threads are assigned with so many functionalities so it is hard to do the consolidations in one shot, we will try to split it into at least two steps (PRs) to make it easier to test and review.

We did our first attemp in https://github.com/pytorch/pytorch/pull/153668 but we also want to try to see if we can make monitoring thread a class. This PR is doing the first step to make monitoring thread a class. The next step to also extract watchdog to be a separate class so that we know its dependency.

What we did in this PR:
1. Move all related variables and methods into a class named `HeartbeatMonitor`.
2. Correct some errors in the original logics inside monitoring thread loop.
3. Move the error propagation check to watchdog thread which is more relevant. This is totally fine since we rolled out EventCache out fully so watchdog hang is rare now.

Today there are two major functions inside heartbeat monitoring thread today: 
1. Check the heartbeat of watchdog thread every 8 minutes. If no heartbeat detected and we are sure monitoring thread has not been stopped, we will kill the program by SIG_ABORT.
2. We check TCPStore every 30 sec to see if any watchdog timeout happens on other ranks, if so we will initiate a dump signal on the current rank as well. (We do this only in the default PG)

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k

Differential Revision: [D75799278](https://our.internmc.facebook.com/intern/diff/D75799278)

fduwjj committed:
[WIP] Consolidate watchdog and monitoring thread

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Consolidate watchdog and monitoring thread"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Consolidate watchdog and monitoring thread"


This is the start of a series of efforts to consolidating auxiliary threads in PGNCCL, aka watchdog and heartbeat_monitoring threads. Right now we launch these two threads per PG instances, i.e., if users create hundred or thousand instances of PG or subPGs, we will end up with that twice many side threads which is not efficient. We have a RFC to consolidate them (https://github.com/pytorch/pytorch/issues/146956). Right now both threads are assigned with so many functionalities so it is hard to do the consolidations in one shot, we will try to split it into at least two steps (PRs) to make it easier to test and review.

We did our first attemp in https://github.com/pytorch/pytorch/pull/153668 but we also want to try to see if we can make monitoring thread a class. This PR is doing similar thing like https://github.com/pytorch/pytorch/pull/153668.

First of all, we start with the heartbeat monitoring thread which is relatively lightweight and conceptually easier to consolidate. What we did in this PR:
1. Make the heartbeat a singleton and make it static, instead of launching it per PGNCCL instance. We make all the logic or variables used in this thread either global or static so that it does not call instance specific API.
2. Remove the dependency on PGStatus which is PG instance specific. (We need to do more around PG Status if we want to consolidate watchdog thread later but it is out of the scope of this PR.)
3. Move the error propagation check to watchdog thread which is more relevant. This is totally fine since we rolled out EventCache out fully so watchdog hang is rare now.

Today there are two major functions inside heartbeat monitoring thread today: 
1. Check the heartbeat of watchdog thread every 8 minutes, we make the heartbeat of watchdog global instead of instance specific. If there are watchdog hang, it should be global. (I am open to better solutions to this one).
2. We check TCPStore every 30 sec to see if any watchdog timeout happens on other ranks, if so we will initiate a dump signal on the current rank as well. 

Previously we only let the thread on the default PG instance to do #2, now with this consolidation, we do FR dump signal every 30 secs and heartbeat check every 8 mins at the same time. If we break the polling loop early we will wait until the full heartbeat timeout (8 mins) before killing the whole program (when we first built heartbeat thread we want to directly kill the program when the watchdog or the whole program hang at CudaEvent destory or NCCL Abort).

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

d4l3k commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1652,36 +1625,117 @@ std::string ProcessGroupNCCL::getNCCLWatchdogTimeoutExitMsg(
+    ProcessGroupNCCL* pg) {
+  pg_ = pg;
+  rank_ = rank;
+  size_ = pg_->getSize();
```
we're going to have to get rid of this init function to support multiple PGs right?

fduwjj replied:
no the idea is to consolidate this thread as well, so it will work even for multiple PG because the default PG will always be created first. That's the assumption we have for globalRank() as well.

d4l3k commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -596,6 +596,116 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+   public:
+    // Return the singleton instance of HeartbeatMonitor.
+    static HeartbeatMonitor* get() {
+      static HeartbeatMonitor instance;
```
nit: should put static in a .cpp file otherwise you could end up with multiple if you reference this function from different CPP files. I assume that's not intentional?

fduwjj replied:
oh I see.. I can move it to cpp file.

d4l3k reviewed (commented):
seems like a good start but definitely can't land this as is right? it'll cause some conflicts when using multiple PGs at the same time if I understand correctly

fduwjj commented:
@d4l3k the call_once and singleton guarantee that it will work even for multiple PG situations.

fduwjj committed:
Update on "[RFC] Consolidate watchdog and monitoring thread"


This is the start of a series of efforts to consolidating auxiliary threads in PGNCCL, aka watchdog and heartbeat_monitoring threads. Right now we launch these two threads per PG instances, i.e., if users create hundred or thousand instances of PG or subPGs, we will end up with that twice many side threads which is not efficient. We have a RFC to consolidate them (https://github.com/pytorch/pytorch/issues/146956). Right now both threads are assigned with so many functionalities so it is hard to do the consolidations in one shot, we will try to split it into at least two steps (PRs) to make it easier to test and review.

We did our first attemp in https://github.com/pytorch/pytorch/pull/153668 but we also want to try to see if we can make monitoring thread a class. This PR is doing similar thing like https://github.com/pytorch/pytorch/pull/153668.

First of all, we start with the heartbeat monitoring thread which is relatively lightweight and conceptually easier to consolidate. What we did in this PR:
1. Make the heartbeat a singleton and make it static, instead of launching it per PGNCCL instance. We make all the logic or variables used in this thread either global or static so that it does not call instance specific API.
2. Remove the dependency on PGStatus which is PG instance specific. (We need to do more around PG Status if we want to consolidate watchdog thread later but it is out of the scope of this PR.)
3. Move the error propagation check to watchdog thread which is more relevant. This is totally fine since we rolled out EventCache out fully so watchdog hang is rare now.

Today there are two major functions inside heartbeat monitoring thread today: 
1. Check the heartbeat of watchdog thread every 8 minutes, we make the heartbeat of watchdog global instead of instance specific. If there are watchdog hang, it should be global. (I am open to better solutions to this one).
2. We check TCPStore every 30 sec to see if any watchdog timeout happens on other ranks, if so we will initiate a dump signal on the current rank as well. 

Previously we only let the thread on the default PG instance to do #2, now with this consolidation, we do FR dump signal every 30 secs and heartbeat check every 8 mins at the same time. If we break the polling loop early we will wait until the full heartbeat timeout (8 mins) before killing the whole program (when we first built heartbeat thread we want to directly kill the program when the watchdog or the whole program hang at CudaEvent destory or NCCL Abort).

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[RFC] Consolidate watchdog and monitoring thread"


This is the start of a series of efforts to consolidating auxiliary threads in PGNCCL, aka watchdog and heartbeat_monitoring threads. Right now we launch these two threads per PG instances, i.e., if users create hundred or thousand instances of PG or subPGs, we will end up with that twice many side threads which is not efficient. We have a RFC to consolidate them (https://github.com/pytorch/pytorch/issues/146956). Right now both threads are assigned with so many functionalities so it is hard to do the consolidations in one shot, we will try to split it into at least two steps (PRs) to make it easier to test and review.

We did our first attemp in https://github.com/pytorch/pytorch/pull/153668 but we also want to try to see if we can make monitoring thread a class. This PR is doing similar thing like https://github.com/pytorch/pytorch/pull/153668.

First of all, we start with the heartbeat monitoring thread which is relatively lightweight and conceptually easier to consolidate. What we did in this PR:
1. Make the heartbeat a singleton and make it static, instead of launching it per PGNCCL instance. We make all the logic or variables used in this thread either global or static so that it does not call instance specific API.
2. Remove the dependency on PGStatus which is PG instance specific. (We need to do more around PG Status if we want to consolidate watchdog thread later but it is out of the scope of this PR.)
3. Move the error propagation check to watchdog thread which is more relevant. This is totally fine since we rolled out EventCache out fully so watchdog hang is rare now.

Today there are two major functions inside heartbeat monitoring thread today: 
1. Check the heartbeat of watchdog thread every 8 minutes, we make the heartbeat of watchdog global instead of instance specific. If there are watchdog hang, it should be global. (I am open to better solutions to this one).
2. We check TCPStore every 30 sec to see if any watchdog timeout happens on other ranks, if so we will initiate a dump signal on the current rank as well. 

Previously we only let the thread on the default PG instance to do #2, now with this consolidation, we do FR dump signal every 30 secs and heartbeat check every 8 mins at the same time. If we break the polling loop early we will wait until the full heartbeat timeout (8 mins) before killing the whole program (when we first built heartbeat thread we want to directly kill the program when the watchdog or the whole program hang at CudaEvent destory or NCCL Abort).

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1624,64 +1590,108 @@ static long computeDeltaMS(
 }
 
-std::string ProcessGroupNCCL::getNCCLWatchdogTimeoutErrorMsg(
+void ProcessGroupNCCL::setEnableNanCheck(bool enableNanCheck) {
+  enableNanCheck_ = enableNanCheck;
+}
```
this is not new added it is just because we move other functions around.

fduwjj committed:
Update on "[RFC] Consolidate watchdog and monitoring thread"


This is the start of a series of efforts to consolidating auxiliary threads in PGNCCL, aka watchdog and heartbeat_monitoring threads. Right now we launch these two threads per PG instances, i.e., if users create hundred or thousand instances of PG or subPGs, we will end up with that twice many side threads which is not efficient. We have a RFC to consolidate them (https://github.com/pytorch/pytorch/issues/146956). Right now both threads are assigned with so many functionalities so it is hard to do the consolidations in one shot, we will try to split it into at least two steps (PRs) to make it easier to test and review.

We did our first attemp in https://github.com/pytorch/pytorch/pull/153668 but we also want to try to see if we can make monitoring thread a class. This PR is doing similar thing like https://github.com/pytorch/pytorch/pull/153668.

First of all, we start with the heartbeat monitoring thread which is relatively lightweight and conceptually easier to consolidate. What we did in this PR:
1. Make the heartbeat a singleton and make it static, instead of launching it per PGNCCL instance. We make all the logic or variables used in this thread either global or static so that it does not call instance specific API.
2. Remove the dependency on PGStatus which is PG instance specific. (We need to do more around PG Status if we want to consolidate watchdog thread later but it is out of the scope of this PR.)
3. Move the error propagation check to watchdog thread which is more relevant. This is totally fine since we rolled out EventCache out fully so watchdog hang is rare now.

Today there are two major functions inside heartbeat monitoring thread today: 
1. Check the heartbeat of watchdog thread every 8 minutes, we make the heartbeat of watchdog global instead of instance specific. If there are watchdog hang, it should be global. (I am open to better solutions to this one).
2. We check TCPStore every 30 sec to see if any watchdog timeout happens on other ranks, if so we will initiate a dump signal on the current rank as well. 

Previously we only let the thread on the default PG instance to do #2, now with this consolidation, we do FR dump signal every 30 secs and heartbeat check every 8 mins at the same time. If we break the polling loop early we will wait until the full heartbeat timeout (8 mins) before killing the whole program (when we first built heartbeat thread we want to directly kill the program when the watchdog or the whole program hang at CudaEvent destory or NCCL Abort).

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2360,16 +2390,14 @@ void ProcessGroupNCCL::watchdogHandler() {
         // try to notify other ranks via global TCPStore to dump the flight
         // recorder when a collective timeout or exception happens. Flight
         // recorder behavior is independent of desync Debug.
-        if (dumpOnTimeoutOrEx_) {
```
we already launched this so we can safely let it broadcast the signal..

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1872,39 +1890,41 @@ void ProcessGroupNCCL::heartbeatMonitor() {
+  //
+  // Or we get stuck in destructors, we will sleep for some time before calling
+  // std::abort() to kill the whole process.
+  if (shouldDump_.load() && !terminateHeartbeatMonitorThread_.load()) {
```
Update the comment and logic here because the original hypothesis is not correct.

fduwjj committed:
Update on "[RFC] Consolidate watchdog and monitoring thread"


This is the start of a series of efforts to consolidating auxiliary threads in PGNCCL, aka watchdog and heartbeat_monitoring threads. Right now we launch these two threads per PG instances, i.e., if users create hundred or thousand instances of PG or subPGs, we will end up with that twice many side threads which is not efficient. We have a RFC to consolidate them (https://github.com/pytorch/pytorch/issues/146956). Right now both threads are assigned with so many functionalities so it is hard to do the consolidations in one shot, we will try to split it into at least two steps (PRs) to make it easier to test and review.

We did our first attemp in https://github.com/pytorch/pytorch/pull/153668 but we also want to try to see if we can make monitoring thread a class. This PR is doing similar thing like https://github.com/pytorch/pytorch/pull/153668.

First of all, we start with the heartbeat monitoring thread which is relatively lightweight and conceptually easier to consolidate. What we did in this PR:
1. Make the heartbeat a singleton and make it static, instead of launching it per PGNCCL instance. We make all the logic or variables used in this thread either global or static so that it does not call instance specific API.
2. Remove the dependency on PGStatus which is PG instance specific. (We need to do more around PG Status if we want to consolidate watchdog thread later but it is out of the scope of this PR.)
3. Move the error propagation check to watchdog thread which is more relevant. This is totally fine since we rolled out EventCache out fully so watchdog hang is rare now.

Today there are two major functions inside heartbeat monitoring thread today: 
1. Check the heartbeat of watchdog thread every 8 minutes, we make the heartbeat of watchdog global instead of instance specific. If there are watchdog hang, it should be global. (I am open to better solutions to this one).
2. We check TCPStore every 30 sec to see if any watchdog timeout happens on other ranks, if so we will initiate a dump signal on the current rank as well. 

Previously we only let the thread on the default PG instance to do #2, now with this consolidation, we do FR dump signal every 30 secs and heartbeat check every 8 mins at the same time. If we break the polling loop early we will wait until the full heartbeat timeout (8 mins) before killing the whole program (when we first built heartbeat thread we want to directly kill the program when the watchdog or the whole program hang at CudaEvent destory or NCCL Abort).

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -596,6 +596,87 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+  // we can dump the debugging information and abort the process.
+  class HeartbeatMonitor {
+   public:
+    HeartbeatMonitor(ProcessGroupNCCL* pg);
```
Use cpp way. `intrusive_ptr` or `shared_ptr`. 
Also, to avoid cyclic dependency problem, this shared ptr can be a weak pointer:
> A weak pointer is a type of smart pointer that allows you to hold a non-owning reference to an object that is owned by a shared pointer.

fduwjj replied:
Actually that's why we don't use intrusive_ptr or shared_ptr here to avoid a cyclic dependency. And a raw pointer is indeed is ok here because the life cycle of this class is always within PGNCCL so we won't have illegal access?

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -596,6 +596,87 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+    // Start the heartbeat monitor thread.
+    void start();
+
+    // Stop the heartbeat monitor thread.
+    void wait();
+
+    // Run the actual loop to check watchdog heartbeat.
+    virtual void runLoop();
+
+    // notify the heartbeat monitor thread to stop.
+    void stop();
```
can you document more clearly the difference between `wait()` and `stop()`? Or maybe clearer naming?

fduwjj replied:
OK saw your comment below, I can rename it to join.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -596,6 +596,87 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+    // notify the heartbeat monitor thread to stop.
+    void stop();
+
+    // print out ENV settings for the heartbeat monitor thread.
+    void printLogMsg();
```
Move this into constructor thus no need for an API?

fduwjj replied:
Ahh ok..

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -596,6 +596,87 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+    std::string getNCCLWatchdogTimeoutExitMsg(const std::string& exitReason);
+
+   protected:
+    // We need to keep a reference to the PG instance so that we can access
+    // the member functions of the PG instance.
+    ProcessGroupNCCL* pg_;
```
weak shared ptr? See above.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -596,6 +596,87 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+
+    // We gate the heartbeat monitor thread so that we can roll it out
+    // gradually.
+    std::atomic<bool> watchdogHeartbeatMonitorEnabled_{};
```
nit: no need for `atomic` here.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -596,6 +596,87 @@ class TORCH_API ProcessGroupNCCL : public Backend {
+    // Mutex to Guard monitorWakeUpCV_
+    std::mutex monitorMutex_;
+
+    // The last update time of watchdog thread.
+    std::chrono::time_point<std::chrono::steady_clock> lastWorkListUpdateTime_;
```
nit: I recommend keeping consistency between naming and comment.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -658,6 +739,8 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 #endif
   }
 
+  uint64_t getWatchdogHeartbt() const;
+
```
nit: should be "protected"?

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -956,33 +955,14 @@ ProcessGroupNCCL::ProcessGroupNCCL(
-  // across different ProcessGroupNCCL(PG) instances. We need to get the
-  // underlying non-PrefixStore for sharing global information shared across
-  // different PGs.
-  PrefixStore* prefixStore = dynamic_cast<PrefixStore*>(store_.get());
-  globalStore_ =
-      prefixStore ? prefixStore->getUnderlyingNonPrefixStore() : store_;
```
where is the assignment of `globalStore_` now?

fduwjj replied:
No assignment now, but inside globleStore()

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1045,15 +1022,11 @@ ProcessGroupNCCL::ProcessGroupNCCL(
-            << ", TORCH_NCCL_LOG_CPP_STACK_ON_UNCLEAN_SHUTDOWN: "
-            << logCppStackOnUncleanShutdown_;
+            << ", TORCH_NCCL_CUDA_EVENT_CACHE: " << cudaEventCacheEnabled_;
+  heartbeatMonitor_ = std::make_unique<HeartbeatMonitor>(this);
+  heartbeatMonitor_->printLogMsg();
```
merge `printLogMsg` into constructor

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2103,6 +2121,11 @@ const int& ProcessGroupNCCL::globalRank() const {
 }
 
+const c10::intrusive_ptr<Store>& ProcessGroupNCCL::globalStore() const {
+  static c10::intrusive_ptr<Store> globalStore = store_;
```
Avoid using `static` if not necessary -- it gets messy in multi-thread scenario.
The original way of being a member variable of ProcessGroupNCCL is better.

fduwjj replied:
This is actually following the implementation of `globalRank()`.

fduwjj replied:
Sharing the same store is indeed expected because we want the default PG to detect signals from other PGs @d4l3k

kwen2501 replied:
I am referring to a case of 1 thread per rank instead of 1 process per rank.
In that case, using the same store across threads (ranks) would be probably wrong.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1643,37 +1613,89 @@ std::string ProcessGroupNCCL::getNCCLWatchdogTimeoutErrorMsg(
+      << logCppStackOnUncleanShutdown_;
+}
+
+void ProcessGroupNCCL::HeartbeatMonitor::stop() {
+  terminateHeartbeatMonitorThread_.store(true);
+  monitorWakeUpCV_.notify_one();
+}
```
mutex protection ?

fduwjj replied:
why is that? This assign is just one direction.

kwen2501 replied:
Then what is your mutex for?

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1643,37 +1613,89 @@ std::string ProcessGroupNCCL::getNCCLWatchdogTimeoutErrorMsg(
+  monitorWakeUpCV_.notify_one();
+}
+
+void ProcessGroupNCCL::HeartbeatMonitor::start() {
+  ncclHeartbeatMonitorThread_ =
+      std::thread(&ProcessGroupNCCL::HeartbeatMonitor::runLoop, this);
+}
```
check for null'ness of `ncclHeartbeatMonitorThread_` before assignment, to avoid second call of start().

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1643,37 +1613,89 @@ std::string ProcessGroupNCCL::getNCCLWatchdogTimeoutErrorMsg(
+      std::thread(&ProcessGroupNCCL::HeartbeatMonitor::runLoop, this);
+}
+
+void ProcessGroupNCCL::HeartbeatMonitor::wait() {
+  if (ncclHeartbeatMonitorThread_.joinable()) {
+    ncclHeartbeatMonitorThread_.join();
+    LOG(INFO) << pg_->logPrefix()
+              << "ProcessGroupNCCL heart beat monitor thread joined.";
+  }
+}
```
by `wait()` do you actually mean `join()`?

fduwjj replied:
ok, maybe I can rename the `wait` to be `join`.

kwen2501 reviewed (approved):
LGTM.

d4l3k reviewed (approved):
seems reasonable to me

fduwjj committed:
Update on "[c10d] Separate monitoring thread into a class in PGNCCL"


This is the start of a series of efforts to consolidating auxiliary threads in PGNCCL, aka watchdog and heartbeat_monitoring threads. Right now we launch these two threads per PG instances, i.e., if users create hundred or thousand instances of PG or subPGs, we will end up with that twice many side threads which is not efficient. We have a RFC to consolidate them (https://github.com/pytorch/pytorch/issues/146956). Right now both threads are assigned with so many functionalities so it is hard to do the consolidations in one shot, we will try to split it into at least two steps (PRs) to make it easier to test and review.

We did our first attemp in https://github.com/pytorch/pytorch/pull/153668 but we also want to try to see if we can make monitoring thread a class. This PR is doing the first step to make monitoring thread a class. The next step to also extract watchdog to be a separate class so that we know its dependency.

What we did in this PR:
1. Move all related variables and methods into a class named `HeartbeatMonitor`.
2. Correct some errors in the original logics inside monitoring thread loop.
3. Move the error propagation check to watchdog thread which is more relevant. This is totally fine since we rolled out EventCache out fully so watchdog hang is rare now.

Today there are two major functions inside heartbeat monitoring thread today: 
1. Check the heartbeat of watchdog thread every 8 minutes. If no heartbeat detected and we are sure monitoring thread has not been stopped, we will kill the program by SIG_ABORT.
2. We check TCPStore every 30 sec to see if any watchdog timeout happens on other ranks, if so we will initiate a dump signal on the current rank as well. (We do this only in the default PG)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge -i

izaitsevfb commented:
@pytorchbot revert -m "breaks lots of jobs internally, safer to revert, see D75628917" -c ghfirst

fduwjj committed:
Update on "[c10d] Separate monitoring thread into a class in PGNCCL"


This is the start of a series of efforts to consolidating auxiliary threads in PGNCCL, aka watchdog and heartbeat_monitoring threads. Right now we launch these two threads per PG instances, i.e., if users create hundred or thousand instances of PG or subPGs, we will end up with that twice many side threads which is not efficient. We have a RFC to consolidate them (https://github.com/pytorch/pytorch/issues/146956). Right now both threads are assigned with so many functionalities so it is hard to do the consolidations in one shot, we will try to split it into at least two steps (PRs) to make it easier to test and review.

We did our first attemp in https://github.com/pytorch/pytorch/pull/153668 but we also want to try to see if we can make monitoring thread a class. This PR is doing the first step to make monitoring thread a class. The next step to also extract watchdog to be a separate class so that we know its dependency.

What we did in this PR:
1. Move all related variables and methods into a class named `HeartbeatMonitor`.
2. Correct some errors in the original logics inside monitoring thread loop.
3. Move the error propagation check to watchdog thread which is more relevant. This is totally fine since we rolled out EventCache out fully so watchdog hang is rare now.

Today there are two major functions inside heartbeat monitoring thread today: 
1. Check the heartbeat of watchdog thread every 8 minutes. If no heartbeat detected and we are sure monitoring thread has not been stopped, we will kill the program by SIG_ABORT.
2. We check TCPStore every 30 sec to see if any watchdog timeout happens on other ranks, if so we will initiate a dump signal on the current rank as well. (We do this only in the default PG)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Separate monitoring thread into a class in PGNCCL"


This is the start of a series of efforts to consolidating auxiliary threads in PGNCCL, aka watchdog and heartbeat_monitoring threads. Right now we launch these two threads per PG instances, i.e., if users create hundred or thousand instances of PG or subPGs, we will end up with that twice many side threads which is not efficient. We have a RFC to consolidate them (https://github.com/pytorch/pytorch/issues/146956). Right now both threads are assigned with so many functionalities so it is hard to do the consolidations in one shot, we will try to split it into at least two steps (PRs) to make it easier to test and review.

We did our first attemp in https://github.com/pytorch/pytorch/pull/153668 but we also want to try to see if we can make monitoring thread a class. This PR is doing the first step to make monitoring thread a class. The next step to also extract watchdog to be a separate class so that we know its dependency.

What we did in this PR:
1. Move all related variables and methods into a class named `HeartbeatMonitor`.
2. Correct some errors in the original logics inside monitoring thread loop.
3. Move the error propagation check to watchdog thread which is more relevant. This is totally fine since we rolled out EventCache out fully so watchdog hang is rare now.

Today there are two major functions inside heartbeat monitoring thread today: 
1. Check the heartbeat of watchdog thread every 8 minutes. If no heartbeat detected and we are sure monitoring thread has not been stopped, we will kill the program by SIG_ABORT.
2. We check TCPStore every 30 sec to see if any watchdog timeout happens on other ranks, if so we will initiate a dump signal on the current rank as well. (We do this only in the default PG)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

Differential Revision: [D75799278](https://our.internmc.facebook.com/intern/diff/D75799278)

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Separate monitoring thread into a class in PGNCCL"


This is the start of a series of efforts to consolidating auxiliary threads in PGNCCL, aka watchdog and heartbeat_monitoring threads. Right now we launch these two threads per PG instances, i.e., if users create hundred or thousand instances of PG or subPGs, we will end up with that twice many side threads which is not efficient. We have a RFC to consolidate them (https://github.com/pytorch/pytorch/issues/146956). Right now both threads are assigned with so many functionalities so it is hard to do the consolidations in one shot, we will try to split it into at least two steps (PRs) to make it easier to test and review.

We did our first attemp in https://github.com/pytorch/pytorch/pull/153668 but we also want to try to see if we can make monitoring thread a class. This PR is doing the first step to make monitoring thread a class. The next step to also extract watchdog to be a separate class so that we know its dependency.

What we did in this PR:
1. Move all related variables and methods into a class named `HeartbeatMonitor`.
2. Correct some errors in the original logics inside monitoring thread loop.
3. Move the error propagation check to watchdog thread which is more relevant. This is totally fine since we rolled out EventCache out fully so watchdog hang is rare now.

Today there are two major functions inside heartbeat monitoring thread today: 
1. Check the heartbeat of watchdog thread every 8 minutes. If no heartbeat detected and we are sure monitoring thread has not been stopped, we will kill the program by SIG_ABORT.
2. We check TCPStore every 30 sec to see if any watchdog timeout happens on other ranks, if so we will initiate a dump signal on the current rank as well. (We do this only in the default PG)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

Differential Revision: [D75799278](https://our.internmc.facebook.com/intern/diff/D75799278)

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Separate monitoring thread into a class in PGNCCL"


This is the start of a series of efforts to consolidating auxiliary threads in PGNCCL, aka watchdog and heartbeat_monitoring threads. Right now we launch these two threads per PG instances, i.e., if users create hundred or thousand instances of PG or subPGs, we will end up with that twice many side threads which is not efficient. We have a RFC to consolidate them (https://github.com/pytorch/pytorch/issues/146956). Right now both threads are assigned with so many functionalities so it is hard to do the consolidations in one shot, we will try to split it into at least two steps (PRs) to make it easier to test and review.

We did our first attemp in https://github.com/pytorch/pytorch/pull/153668 but we also want to try to see if we can make monitoring thread a class. This PR is doing the first step to make monitoring thread a class. The next step to also extract watchdog to be a separate class so that we know its dependency.

What we did in this PR:
1. Move all related variables and methods into a class named `HeartbeatMonitor`.
2. Correct some errors in the original logics inside monitoring thread loop.
3. Move the error propagation check to watchdog thread which is more relevant. This is totally fine since we rolled out EventCache out fully so watchdog hang is rare now.

Today there are two major functions inside heartbeat monitoring thread today: 
1. Check the heartbeat of watchdog thread every 8 minutes. If no heartbeat detected and we are sure monitoring thread has not been stopped, we will kill the program by SIG_ABORT.
2. We check TCPStore every 30 sec to see if any watchdog timeout happens on other ranks, if so we will initiate a dump signal on the current rank as well. (We do this only in the default PG)

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k

Differential Revision: [D75799278](https://our.internmc.facebook.com/intern/diff/D75799278)

[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D75799278).

fduwjj commented:
Import into fbcode this time and all CI tests passed I will merge this PR again.

fduwjj commented:
@pytorchbot merge -i
